### DISCUSSION TITLE
Directly using Nginx as a server without Uvicorn

### DISCUSSION THREAD

**Discussion by bymoye** [Neoteroi/BlackSheep#205] [General]
hi~
Maybe can try to use [nginx unit](unit.nginx.org) as ASGI server, it is faster than uvicorn and hypercorn.
The configuration file of nginx unit is also quite simple.
```
{
    "listeners": {
        "*:80": {
            "pass": "applications/blacksheep"
        }
    },
 
    "applications": {
        "blacksheep": {
            "type": "python 3.9",
            "processes": 2,
            "path": "/path/app",
            "working_directory":"/path/app",
            "module": "main",
            "callable": "app"
        }
    }
}
```

**Comment by RobertoPrevato** [comment]
Hi @bymoye,
At the moment this would not be possible. Because the first version of BlackSheep, the one that is still benchmarked in TechEmpower benchmarks (I didn't have time to prepare a new PR there), included an implementation of HTTP Server (you can see this at this [tag](https://github.com/Neoteroi/BlackSheep/tree/before-asgi)).

Since I adopted `ASGI`, the library is a pure web framework without the implementation of an HTTP Server. I am planning to eventually re-implement a built-in HTTP Server as I did for the first version linked above (_in addition_ to ASGI support), and I have ideas to make it work faster, but this won't happen soon.

However, using NGINX as proxy server in front of `uvicorn` (HTTP Server) and using `blacksheep` (web framework) still makes sense.

Recently I preferred to dedicate time to work on higher level features rather than putting lots of energies on low level details such as HTTP Server implementation. By the way, right in these days I worked to add built-in support for OpenID Connect and JWT Bearer authentication, features that make it extremely easy to integrate with identity providers such as Auth0, Okta, Azure Active Directory, Azure Active Directory B2C. I plan to release `1.2.1` this weekend with these new features, plus:
* middlewares to handle X-Forwarded and Forwarded headers, and trusted hosts
* JWT Bearer authentication
* OpenID Connect + Cookie authentication

**Reply by bymoye** [reply]
Thanks Reply!

Awesome, thanks for adding these! These are some very useful features. Looking forward to blacksheep getting better and better!
